Prevent duplicate app windows

jansav opened the pull request:
Note: base-branch of this PR is https://github.com/lensapp/lens/pull/5433 instead of master, because the requirements of unit testing are introduced in the branch.

Note: this fixes bug about multiple application windows opening if user reopens Lens using tray while the splash screen was still visible. The bug was reproduced in a unit test and some behaviours around starting and stopping Lens where created.

Relevant test:
opening application window using tray › given application has started › when the application window is closed › when an application window is reopened using tray › given opening of splash window has not finished yet, but another attempt to open the application is made › does not open any new windows

Iku-turso committed:
Rename interface for honesty

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>

Signed-off-by: Iku-turso <mikko.aspiala@gmail.com>
Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

Iku-turso committed:
Add behaviours for quitting the application using application menu

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>

Signed-off-by: Iku-turso <mikko.aspiala@gmail.com>

Iku-turso committed:
Add behaviours for opening application window using tray and fix indicative bug about opening involuntary duplicate app windows

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>

Signed-off-by: Iku-turso <mikko.aspiala@gmail.com>

Iku-turso committed:
Make sure bug about opening involuntary duplicate app windows is solved in all scenarios

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>

Signed-off-by: Iku-turso <mikko.aspiala@gmail.com>
Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

Nokel81 commented on src/behaviours/quitting-and-restarting-the-app/opening-application-window-using-tray.test.ts:
```diff
@@ -0,0 +1,163 @@
+          it("still only splash window is open", () => {
+            expectWindowsToBeOpen(["splash"]);
+          });
+
```
I would like to see another test along the lines of "given another attempt to open the application is made > shows window immediately without waiting for it to load".

I believe that is what the previous behaviour was and is useful when you get a crash before the window has reported that it is loaded, that way we can open the dev tools and look at the logs.

jansav replied:
Hm, sounds definitely like something that we should still have.

Nokel81 replied:
Doesn't look like it has been added yet...

jansav replied:
Yes, it hasn't. We'll be working on this tomorrow.

Iku-turso replied:
Added, and then some.

Nokel81 commented on src/main/start-main-application/lens-window/application-window/create-lens-window.injectable.ts:
```diff
@@ -35,31 +35,45 @@ const createLensWindowInjectable = getInjectable({
           return !!browserWindow;
         },
+
+        get opening() {
```
nit: isOpening might read a bit better.

Iku-turso replied:
Right you are, renamed. Also instead of "opening" there's now a distinction between "starting" and "showing" a window. Therefore this naming is `isStarting`.

ixrock commented on src/behaviours/quitting-and-restarting-the-app/opening-application-window-using-tray.test.ts:
```diff
@@ -0,0 +1,163 @@
+import { lensWindowInjectionToken } from "../../main/start-main-application/lens-window/application-window/lens-window-injection-token";
+import applicationWindowInjectable from "../../main/start-main-application/lens-window/application-window/application-window.injectable";
+import createElectronWindowForInjectable from "../../main/start-main-application/lens-window/application-window/create-electron-window-for.injectable";
+import type { AsyncFnMock } from "@async-fn/jest";
+import asyncFn from "@async-fn/jest";
```
_nit:_ 

```suggestion
import asyncFn, { type AsyncFnMock } from "@async-fn/jest";
```

jansav replied:
We're using auto import + eslint which makes two separate lines. So if we want to switch to using suggested, let's update eslint to do the work for us. :)

Iku-turso committed:
Make loading contents of a window unit testable

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>

Signed-off-by: Iku-turso <mikko.aspiala@gmail.com>
Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Implement showing of blank application window when re-opening before contents have been load to permit developer tool access

Co-authored-by: Mikko Aspiala <mikko.aspiala@gmail.com>

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

Iku-turso committed:
Consolidate naming for less ambiguity

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>

Signed-off-by: Iku-turso <mikko.aspiala@gmail.com>

# Conflicts:
#	src/behaviours/quitting-and-restarting-the-app/opening-application-window-using-tray.test.ts
#	src/main/start-main-application/lens-window/application-window/create-lens-window.injectable.ts

Iku-turso committed:
Make existing tests green by adapting to newly introduced difference between started and shown window

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>

Signed-off-by: Iku-turso <mikko.aspiala@gmail.com>

# Conflicts:
#	src/behaviours/quitting-and-restarting-the-app/opening-application-window-using-tray.test.ts
#	src/main/start-main-application/lens-window/application-window/create-lens-window.injectable.ts

Iku-turso committed:
Add stub value to proxy port to make unrelated tests pass

Note: the actual mechanism to set a value for the port is not in scope of behavioral tests yet.

Co-authored-by: Janne Savolainen <janne.savolainen@live.fi>

Signed-off-by: Iku-turso <mikko.aspiala@gmail.com>

# Conflicts:
#	src/behaviours/quitting-and-restarting-the-app/opening-application-window-using-tray.test.ts
#	src/main/start-main-application/lens-window/application-window/create-lens-window.injectable.ts

Nokel81 commented on src/common/utils/channel/channel.test.ts:
```diff
@@ -94,12 +94,12 @@ describe("channel", () => {
     });
 
-    it("given multiple shown windows, when sending message, triggers listeners in all windows", async () => {
+    it("given multiple started started windows, when sending message, triggers listeners in all windows", async () => {
```
nit: duplicate wording

jansav replied:
Good catch, fixed.

Nokel81 commented on src/main/start-main-application/lens-window/application-window/wait-until-bundled-extensions-are-loaded.injectable.ts:
```diff
@@ -0,0 +1,29 @@
+      });
+
+      await viewHasLoaded;
+      await delay(50); // wait just a bit longer to let the first round of rendering happen
```
Sometime in the future we should add a `requestAnimationFrame` wrapper into the renderer so that this delay is not needed.

jansav replied:
Absolutely. Here we just extracted this to separate injectable so that we can override it from unit tests.

Nokel81 commented on src/main/electron-app/runnables/dock-visibility/hide-dock-for-last-closed-window.injectable.ts:
```diff
@@ -20,7 +20,7 @@ const hideDockForLastClosedWindowInjectable = getInjectable({
           getLensWindows(),
-          filter(window => !!window.visible),
+          filter(window => !!window.isVisible),
         );
```
This code has been duplicated in several places, it probably should be its own injectable.

jansav replied:
Sure. Normally we've done it when same thing is duplicated 3-4 times. But this is clearly case in which we don't have to wait since we know already it's not going to be wrong abstraction. :) Done.

Nokel81 reviewed (commented):
Looks great, just some minor things.

jansav committed:
Fix typo in name of test

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Remove duplication for getting visible windows

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Remove redundant kludge for waiting till next tick for not being needed anymore in jest 28

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

Nokel81 commented on src/main/utils/channel/message-to-channel.injectable.ts:
```diff
@@ -2,32 +2,24 @@
-    const getAllLensWindows = () => di.injectMany(lensWindowInjectionToken);
+    const getVisibleWindows = di.inject(getVisibleWindowsInjectable);
 
     // TODO: Figure out way to improve typing in internals
```
This can be sort of fixed by 

1. Wrap the return value in `(...)` and then add an `as MessageToChannel` cast.

Also, I think it would be possible to simply the type `MessageToChannel` to the following:

```ts
export interface MessageToChannel {
  (channel: MessageChannel<void>): void;
  <Message>(channel: MessageChannel<Message>, message: Message): void;
}
```

jim-docker reviewed (approved):
It's odd that, when clicking on "Open Lens" from the tray shows the splash window and doesn't show the app window until it's "ready", but clicking on "Open Lens" from the tray a second time, while the splash window is still shown, the app window is shown immediately, and it is empty until it becomes "ready". Would be cleaner if the app window does not show until it is "ready", regardless of how many times "Open Lens" is clicked.
